File: 1658-minimum-operations-to-reduce-x-to-zero/1658-minimum-operations-to-reduce-x-to-zero.java
class Solution {
    public int minOperations(int[] nums, int x) {
        int total = 0;
        int i=0;
        boolean found = false;
        int final_ans = nums.length;
        
        while(i<nums.length){
            if (total + nums[i] > x){
                i-=1;
                break;
            }
            else if (total + nums[i] == x){
                found = true;
                total += nums[i];
                final_ans = Math.min(final_ans, i+1);
                break;
            }
            total += nums[i];
            i+=1;
        }
        
        
        if (i==nums.length){
            return !found ? -1 : final_ans;
        } 
        
        // System.out.println(i + " " + final_ans);
        for (int j=nums.length-1; j>-1; j--){
            while (i>-1 && total + nums[j] > x){
                total -= nums[i]; 
                i-=1;
            }
            
            total += nums[j];
            if (total == x){
                found=true;
                final_ans = Math.min(final_ans, nums.length-j+i+1);
            }
            else if (total > x){
                return !found ? -1 : final_ans;
            }
            // System.out.println(total + " " + i + " " + j + " " + final_ans);
                        
        }
        return !found ? -1 : final_ans;
    }
}

File: 1658-minimum-operations-to-reduce-x-to-zero/1658-minimum-operations-to-reduce-x-to-zero.py
class Solution:
    def minOperations(self, nums: List[int], x: int) -> int:
        '''
        Input: nums = [1,1,4,2,3], x = 5
        
        0,1,2,6,8,11
        '''
        p_sum = [0]
        for num in nums:
            p_sum.append(num + p_sum[-1])
        
        ans = len(nums) + 1
        
        right_sum = 0
        
        for right in range(len(p_sum)-1, -1, -1):
            
            l, r = 0, right
            while l <= r:
                mid = l + (r-l)//2
                
                if right_sum + p_sum[mid] == x:
                    ans = min(ans, mid + len(p_sum)-right-1 )
                    break
                    
                if right_sum + p_sum[mid] < x:
                    l = mid+1
                else:
                    r = mid-1
            
            right_sum += nums[right-1]
        
        if ans == len(nums) + 1:
            return -1
        return ans